src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs

```diff
@@ -2269,75 +2277,21 @@ private static bool IsThisOrTypeOrNamespace(MemberAccessExpressionSyntax memberA
             OptionSet optionSet,
             CancellationToken cancellationToken)
         {
-            replacementNode = null;
-            issueSpan = default;
+            var typeStyle = CSharpUseImplicitTypeHelper.Instance.AnalyzeTypeName(
+                simpleName, semanticModel, optionSet, cancellationToken);
 
-            if (!optionSet.GetOption(SimplificationOptions.PreferImplicitTypeInLocalDeclaration))
+            if (!typeStyle.IsStylePreferred || !typeStyle.CanConvert())
             {
+                replacementNode = null;
+                issueSpan = default;
                 return false;
             }
 
-            // If it is already var
-            if (simpleName.IsVar)
-            {
-                return false;
-            }
-
-            var candidateReplacementNode = SyntaxFactory.IdentifierName("var")
-                                            .WithLeadingTrivia(simpleName.GetLeadingTrivia())
-                                            .WithTrailingTrivia(simpleName.GetTrailingTrivia());
-            var candidateIssueSpan = simpleName.Span;
-
-            // If there exists a Type called var , fail.
-            var checkSymbol = semanticModel.GetSpeculativeSymbolInfo(simpleName.SpanStart, candidateReplacementNode, SpeculativeBindingOption.BindAsTypeOrNamespace).Symbol;
-            if (checkSymbol != null && checkSymbol.IsKind(SymbolKind.NamedType) && ((INamedTypeSymbol)checkSymbol).TypeKind == TypeKind.Class && checkSymbol.Name == "var")
-            {
-                return false;
-            }
-
-            // If the simpleName is the type of the Variable Declaration Syntax belonging to LocalDeclaration, For Statement or Using statement
-            if (simpleName.IsParentKind(SyntaxKind.VariableDeclaration) &&
```

Comment:
we had duplicated a bunch of "can i use var" logic here.  It was both broken and incomplete. Now, we just defer to the CSharpUseImplicitTypeHelper which contains all the correct logic and is the place we should always be updating.